File: sdks/php/rest/src/Channels/Requests/PublishRequest.php
<?php

namespace Brume\Rest\Channels\Requests;

use Brume\Rest\Core\Json\JsonSerializableType;
use Brume\Rest\Core\Json\JsonProperty;
use Brume\Rest\Core\Types\ArrayType;

class PublishRequest extends JsonSerializableType
{
    /**
     * Event name. 1-256 chars. Allowed: alphanumeric, `:`, `-`, `_`, `.`.
     * Must NOT start with `brume:` (reserved for system events).
     *
     * @var string $event
     */
    #[JsonProperty('event')]
    public string $event;

    /**
     * @var array<string, mixed> $payload Free-form JSON payload delivered to subscribers.
     */
    #[JsonProperty('payload'), ArrayType(['string' => 'mixed'])]
    public array $payload;

    /**
     * @var ?string $ref Optional client-generated idempotency key. Echoed in `brume:ack`.
     */
    #[JsonProperty('ref')]
    public ?string $ref;

    /**
     * @param array{
     *   event: string,
     *   payload: array<string, mixed>,
     *   ref?: ?string,
     * } $values
     */
    public function __construct(
        array $values,
    ) {
        $this->event = $values['event'];
        $this->payload = $values['payload'];
        $this->ref = $values['ref'] ?? null;
    }
}

File: sdks/csharp/rest/src/Brume.Rest/Channels/Requests/PublishRequest.cs
using Brume.Rest.Core;
using global::System.Text.Json.Serialization;

namespace Brume.Rest;

[Serializable]
public record PublishRequest
{
    /// <summary>
    /// Event name. 1-256 chars. Allowed: alphanumeric, `:`, `-`, `_`, `.`.
    /// Must NOT start with `brume:` (reserved for system events).
    /// </summary>
    [JsonPropertyName("event")]
    public required string Event { get; set; }

    /// <summary>
    /// Free-form JSON payload delivered to subscribers.
    /// </summary>
    [JsonPropertyName("payload")]
    public Dictionary<string, object?> Payload { get; set; } = new Dictionary<string, object?>();

    /// <summary>
    /// Optional client-generated idempotency key. Echoed in `brume:ack`.
    /// </summary>
    [JsonPropertyName("ref")]
    public string? Ref { get; set; }

    /// <inheritdoc />
    public override string ToString()
    {
        return JsonUtils.Serialize(this);
    }
}
